src/test/java/unit/FineGrainedAccessGcsObjectReference.java

```diff
@@ -0,0 +1,462 @@
+package unit;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static unit.GcsObjectReferenced.listObjectResourceWithName;
+
+import bio.terra.cli.serialization.userfacing.UFWorkspace;
+import bio.terra.cli.serialization.userfacing.resource.UFGcsObject;
+import com.google.cloud.Identity;
+import com.google.cloud.storage.Acl;
+import com.google.cloud.storage.Acl.Role;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.Storage;
+import harness.TestCommand;
+import harness.TestUsers;
+import harness.baseclasses.SingleWorkspaceUnit;
+import harness.utils.Auth;
+import harness.utils.ExternalGCSBuckets;
+import java.io.IOException;
+import java.util.List;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+/** Test for CLI command on GCS objects in a fine-grained access GCS bucket. */
+@Tag("unit")
+public class FineGrainedAccessGcsObjectReference extends SingleWorkspaceUnit {
+
+  // external bucket to use for creating GCS bucket references in the workspace
+  private BucketInfo externalBucket;
+
+  // name of blobs in external bucket
+  private String sharedExternalBlobName = "foo/text.txt";
+  private String privateExternalBlobName = "foo/";
+
+  private TestUsers shareeUser;
+  private Blob sharedBlob;
+
+  @BeforeAll
+  @Override
+  protected void setupOnce() throws Exception {
+    super.setupOnce();
+    externalBucket = ExternalGCSBuckets.createBucketWithFineGrainedAccess();
+
+    String proxyGroupEmail = Auth.getProxyGroupEmail();
+
+    // grant the user's proxy group access to the bucket so that it will pass WSM's access check
+    // when adding it as a referenced resource
+    ExternalGCSBuckets.grantWriteAccess(externalBucket, Identity.group(proxyGroupEmail));
+
+    sharedBlob =
```

Comment:
ok so you have the case where the user has permission only on the blob, and the case where the user has permission on the whole bucket. both of those are good test cases. but what I was trying to say in the comment was to check the case where the user has permission on the folder, but not the whole bucket. e.g. write a single new blob "foo2/secret.txt", grant the `shareeUser` permission on "foo2/" only, then check that the `shareeUser` can add a new reference to "foo2/".

Reply:
i actually can't grant access to the foo2/ bucket without creating it. 

When I was adding the integration test in WSM, I used the console UI to create the folder and added object in the folder and then grant distinct access to the folder and the object.

Reply:
sorry, not completely following. If there is just a folder, and no object (e.g. "gs://mybucket/foo2/secret.txt" is the only object, folder is "foo2"), is it possible to grant access to just that folder, either through the console UI or Java client library? or do you always have to grant access to an object?

Reply:
folder is an object. so if i create a foo2/ folder, i can grant access to foo2/. The console uI doesn't let you grant access to a folder but you can do that through gsutil in command line. 

the issue here is that if I programmatically created a foo2/bar.txt object, cloud storage doesn't automatically create a foo2/ object as well. so if I try to grant access to foo2/, this object does not exist.

Reply:
can you clarify what you mean by "create a foo2/" folder? do you mean to write a non-zero size blob to "gs://mybucket/foo2/"?

Reply:
It can be zero-bite.
gcloud documentation: https://cloud.google.com/storage/docs/folders#overview

"If you create an empty folder using the Cloud Console, Cloud Storage creates a zero-byte object as a placeholder. For example, if you create a folder called folder in a bucket called my-bucket, a zero- byte object called gs://my-bucket/folder/ is created. You can find this object when you run gsutil ls."